### PR TITLE
Fix setting preferred color scheme logic for brave theme change

### PR DESCRIPTION
**Pull request by simonhong**
## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [x] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: browser/themes/brave_theme_utils.cc [hunk 1]
```diff
@@ -15,28 +16,15 @@ namespace ui {
 // static
 void BraveThemeUtils::SetDarkMode(bool dark_mode) {
   NativeTheme::GetInstanceForNativeUi()->set_dark_mode(dark_mode);
+  NativeTheme::GetInstanceForWeb()->set_dark_mode(dark_mode);
 }
 
 // static
-void BraveThemeUtils::SetPreferredColorScheme(BraveThemeType brave_theme_type) {
```
Comment (simonhong) [inline browser/themes/brave_theme_utils.cc]: For re-calculating color scheme, I think we just need to set dark mode in native theme before calling `CalculatePreferredColorScheme`.

File: patches/chrome-browser-chrome_content_browser_client.cc.patch [hunk 2]
```diff
@@ -1,13 +0,0 @@
-diff --git a/chrome/browser/chrome_content_browser_client.cc b/chrome/browser/chrome_content_browser_client.cc
-index 0b29bc11f07c95f6048c9a35150925ad16a08586..44cab1c7f2897fb0191b33653d3cae945234d79f 100644
---- a/chrome/browser/chrome_content_browser_client.cc
-+++ b/chrome/browser/chrome_content_browser_client.cc
-@@ -3376,7 +3376,7 @@ void ChromeContentBrowserClient::OverrideWebkitPrefs(
-                                  : blink::ForcedColors::kNone;
- #endif  // !defined(OS_MACOSX)
- 
--  switch (native_theme->GetPreferredColorScheme()) {
```
Comment (simonhong) [inline patches/chrome-browser-chrome_content_browser_client.cc.patch]: I think `GetPreferredColorScheme` will return proper scheme when we set our dark mode and re-calculateded color scheme.

File: patches/ui-native_theme-native_theme_win.h.patch [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
-index 03be75f99e3da8c29a8bc856da9e6f6ceea9af20..2b6aa17d081c90eadca6976a1370d32ace3250d7 100644
+index 03be75f99e3da8c29a8bc856da9e6f6ceea9af20..96221ee7f855cbb246d6427b17b95517c6894032 100644
 --- a/ui/native_theme/native_theme_win.h
 +++ b/ui/native_theme/native_theme_win.h
-@@ -85,6 +85,7 @@ class NATIVE_THEME_EXPORT NativeThemeWin : public NativeTheme,
+@@ -84,6 +84,7 @@ class NATIVE_THEME_EXPORT NativeThemeWin : public NativeTheme,
+   bool SystemDarkModeSupported() const override;
    PreferredColorScheme CalculatePreferredColorScheme() const override;
  
++  BRAVE_UI_NATIVE_THEME_NATIVE_THEME_WIN_H_
```
Comment (simonhong) [inline patches/ui-native_theme-native_theme_win.h.patch]: just clean up.

File: browser/themes/brave_theme_utils.cc [hunk 4]
```diff
@@ -15,28 +16,15 @@ namespace ui {
...
-    default:
-      NOTREACHED();
-  }
-  NativeTheme::GetInstanceForNativeUi()->set_preferred_color_scheme(
-      preferred_color_scheme);
+void BraveThemeUtils::ReCalcAndSetPreferredColorScheme() {
+  auto scheme =
+      NativeTheme::GetInstanceForNativeUi()->CalculatePreferredColorScheme();
+  NativeTheme::GetInstanceForNativeUi()->set_preferred_color_scheme(scheme);
+  NativeTheme::GetInstanceForWeb()->set_preferred_color_scheme(scheme);
```
Comment (simonhong) [inline browser/themes/brave_theme_utils.cc]: W/o considering web, non dark mode platform (ex, linux) doesn't change web dark mode by changing brave theme.

**Comment by mkarolin** [comment]
Unless I have wrong expectations, this is where it fails for me:

1. On Windows, go to Settings -> Personalization -> Colors and set "Choose your color" to Dark.
2. Start Brave, go to brave://settings -> Appearance
3. By default "Brave colors" is set to "Same as Windows" and the UI looks Dark.
4. Switch "Brave colors" to Dark (expect no change in the UI).
5. Go back to Windows -> Settings -> Personalization -> Colors and set "Choose your color" to Light.
6. Expectation: Brave UI remains Dark
    Actual: The window chrome color remains Dark, but the settings page changes to Light.

Same in the opposite direction (Brave color set to Light, windows color switched to Dark => results in the settings page switching to Dark).

cc: @bsclifton - I had a chance to test, so you don't have to.

**Comment by bsclifton** [comment]
Thanks for testing, @mkarolin! 😄 

I can confirm your findings on Windows - seems the Chromium pages are not respecting the custom `Dark` setting ☹️ 

On macOS, it does work as expected (stays Dark after changing macOS to light theme) 👍

**Comment by simonhong** [comment]
Thanks for checking @mkarolin @bsclifton !
I confirmed that this regression is introduced by https://github.com/brave/brave-core/pull/2905 since `0.69.x`. System change can affect brave theme option now. I'm looking at this.

**Comment by simonhong** [comment]
@mkarolin @bsclifton Pushed one more commits for fixing above issue. https://github.com/brave/brave-core/pull/3281/commits/fd2565b833e146ca44fdefe1117ff0ec804df9d2

**Comment by mkarolin** [comment]
@simonhong Thanks! Checked on Windows and it works as expected.

**Comment by simonhong** [comment]
@mkarolin Thanks for checking. I'm trying to add some additional browser tests.
Will let you know when I finish it.

**Comment by simonhong** [comment]
@mkarolin @bsclifton Tests are added and verified all `BraveThemeService*Test` are passed on all platforms.

**Comment by mkarolin** [comment]
It might be easier to cherry-pick these commits into c77 branch directly since it was force pushed and a rebase maybe cumbersome.

**Comment by mkarolin** [comment]
Cherry-picked directly into `77.0.3865.42` branch (#3247).

**Pull request closed without merging**